openapi_metadata.go

```
@@ -0,0 +1,449 @@
+func (mp *MetadataParsers) Initialize(customParsers []MetadataParserEntry) {
+	mp.parserLock.Lock()
+	defer mp.parserLock.Unlock()
+
+	mp.registeredParsers = []MetadataParserEntry{}
+	mp.registeredNames = make(map[string]bool)
+
+	var parsersToRegister = customParsers
+
+	for _, entry := range parsersToRegister {
+		if _, exists := mp.registeredNames[entry.Name]; exists {
+			slog.Warn("Parser already registered", "name", entry.Name)
+			continue
+		}
+		mp.registeredNames[entry.Name] = true
+		mp.registeredParsers = append(mp.registeredParsers, entry)
+	}
+}
+
+// RegisterMetadataParser registers a new metadata parser with the given name, parser
+// function, and position string. The position string can be one of the following:
+// "prepend", "append", "before", or "after". If the position is "before" or "after", you
+// must also provide the name of the parser that you want to place the new parser
+// before or after. If the relative parser is not found, an error is returned. If the
+// parser is already registered, the function does nothing and returns nil. The
+// function is thread-safe and ensures that the registeredParsers slice and
+// registeredNames map are accessed in a thread-safe manner.
+func (mp *MetadataParsers) RegisterMetadataParser(name string, parser MetadataParserFunction, position string, relativeTo string) error {
+	validPositions := map[string]bool{"prepend": true, "append": true, "before": true, "after": true}
+	if !validPositions[position] {
+		return fmt.Errorf("Invalid position. Use 'prepend', 'append', 'before', or 'after'")
+	}
+	if (position == "before" || position == "after") && mp.findParserIndex(relativeTo) == -1 {
+		return fmt.Errorf("Relative parser '%s' not found", relativeTo)
+	}
+
+	mp.parserLock.Lock()
+	defer mp.parserLock.Unlock()
+
+	for _, entry := range mp.registeredParsers {
+		if entry.Name == name {
+			// Remove the existing parser before re-registering
+			mp.remove(entry)
```

Error is not checked.
Please run `make` to check the CI